lightning/src/blinded_path/payment.rs

```diff
@@ -0,0 +1,285 @@
+}
+
+impl Writeable for BlindedPaymentTlvs {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		// TODO: write padding
+		match self {
+			Self::Forward { short_channel_id, payment_relay, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(2, short_channel_id, required),
+					(10, payment_relay, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			},
+			Self::Receive { payment_secret, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(6, payment_secret, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for BlindedPaymentTlvs {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		_init_and_decode_tlv_stream!(r, {
+			(1, _padding, option),
+			(2, scid, option),
+			(6, payment_secret, option),
+			(10, payment_relay, option),
+			(12, payment_constraints, required),
+			(14, features, required),
+		});
+		if let Some(short_channel_id) = scid {
+			if payment_secret.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Forward {
+				short_channel_id,
+				payment_relay: payment_relay.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		} else {
+			if payment_relay.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Receive {
+				payment_secret: payment_secret.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		}
+	}
+}
+
+/// Construct blinded payment hops for the given `unblinded_path`.
+pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
+	secp_ctx: &Secp256k1<T>, unblinded_path: &[(PublicKey, BlindedPaymentTlvs)], session_priv: &SecretKey
+) -> Result<Vec<BlindedHop>, secp256k1::Error> {
+	let mut blinded_hops = Vec::with_capacity(unblinded_path.len());
+	let mut curr_hop_idx = 0;
+	utils::construct_keys_callback(
+		secp_ctx, unblinded_path.iter().map(|(pk, _)| pk), None, session_priv,
+		|blinded_node_id, _, _, encrypted_payload_ss, _, _| {
+			blinded_hops.push(BlindedHop {
+				blinded_node_id,
+				encrypted_payload: utils::encrypt_payload(&unblinded_path[curr_hop_idx].1, encrypted_payload_ss),
+			});
+			curr_hop_idx += 1;
+		})?;
+	Ok(blinded_hops)
+}
+
+pub(super) fn compute_payinfo(
+	path: &[(PublicKey, BlindedPaymentTlvs)]
+) -> Result<BlindedPayInfo, ()> {
+	let mut curr_base_fee: u128 = 0;
+	let mut curr_prop_mil: u128 = 0;
+	for (_, payment_tlvs) in path.iter().rev().skip(1) {
+		let next_base_fee = payment_tlvs.fee_base_msat() as u128;
+		let next_prop_mil = payment_tlvs.fee_proportional_millionths() as u128;
+		curr_base_fee =
+			((next_base_fee * 1_000_000 + (curr_base_fee * (1_000_000 + next_prop_mil))) + 1_000_000 - 1)
+			 / 1_000_000;
```

Comment:
I stared at this for like 10 minutes before realizing this is the same as
```rust
curr_base_fee *= 1 + next_prop_mil / 1_000_000;
curr_base_fee += next_base_fee;
```
but scaling by 1 million for integer math. Anyway, looks good (and identical to the spec)!